packages/kbn-cli-dev-mode/src/base_path_proxy/http2.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Url from 'url';
```

Comment:
- http1 waits and logs `please hold  Kibana server is not ready so requests have been paused`
- http2 closes the conection

Do we want these to match?  Can be reproduced by dropping the bundle cache and loading localhost:5601

Reply:
How much effort would it be to add an http -> https redirect if http2 is used?

Reply:
> http1 waits and logs please hold  Kibana server is not ready so requests have been paused

Hum, I'm surprised this doesn't work with the new implem, given this wait is done outside of the proxy itself

https://github.com/elastic/kibana/blob/41b7951ac691e6782c2569abe473a06750b4ca01/packages/kbn-cli-dev-mode/src/cli_dev_mode.ts#L207-L213

and the http2 proxy waits it for it too:

https://github.com/elastic/kibana/blob/fd5b5a8b5c24af48d365d7e2757d83c13e005773/packages/kbn-cli-dev-mode/src/base_path_proxy/http2.ts#L134-L136


> ~http2 closes the conection~ just kidding, wrong protocol
> How much effort would it be to add an http -> https redirect if http2 is used?

Yeah, I faced the same problem hitting the server with `http://` the first time, before realizing I need to hit on `https` instead, as you probably did...

The thing is, you cannot easily add an http->https redirect **on the same port**. It's not supported by the node listeners.

What I could do though, is leveraging / mimicking  `ssl.redirectHttpFromPort` option, so that when `--http2` is used, we open two port: we listen to `5601` on `http` (http1), `5602`  on `https` (http2), and having `5601` redirect to `5602`.

Would that make sense to you? And if I do that, should I do it for the `--ssl` option too (because atm, using `--ssl` causes the same thing - we listen on https only, without redirecting or anything)

Reply:
> Would you mind checking again?

Looks like I was using the wrong protocol there too, LGTM.  Sorry about that.

Holding off the http->https redirect seems reasonable. It's fairly involved for a non-default workflow, with caveats.  --ssl has existed for a while and it's opt-in similarly, and I haven't heard anyone having issues with it.